### ISSUE TITLE
RFC: Ufunc return type for zero dimensional array input.

### ISSUE THREAD

**Issue by steppi** [scipy/scipy#18768]
## Introduction

(skip to TLDR if you don't care to read this)

The longstanding behavior for NumPy ufuncs has been to return a scalar in cases where the output would otherwise be expected to be a zero dimensional array. This is accomplished in the NumPy C API with the function [PyArray_Return](https://github.com/numpy/numpy/blob/85385d6eb9da2b45d1e5fa9994e82608e562ef32/numpy/core/src/multiarray/scalarapi.c#L770C1-L779C34).

```c

/* Return Array Scalar if 0-d array object is encountered */


/*NUMPY_API
 *
 * Return either an array or the appropriate Python object if the array
 * is 0d and matches a Python type.
 * steals reference to mp
 */
NPY_NO_EXPORT PyObject *
PyArray_Return(PyArrayObject *mp)
```

ufuncs in `special` generated with the standard infrastructure [_generate_pxd.py](https://github.com/scipy/scipy/blob/main/scipy/special/_generate_pyx.py) will automatically have this behavior.

There are however functions in special which have ufunc like behavior but are not generated with the standard infrastructure. These include the factorial functions [factorial](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorial.html#scipy.special.factorial), [factorial2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorial2.html#scipy.special.factorial2), and [factorialk](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorialk.html#scipy.special.factorialk). These functions have an `exact=True` option where exact calculations using Python's arbitrary-precision integers can be made for array-like inputs, returning an `object` array of Python `int`s.

Until recently, factorial functions in SciPy were inconsistent with each other in terms of API and behavior. @h-vetinari made a heroic effort to resolve these inconsistencies, discussion took place in issue gh-15600, and the work was accomplished over a number of PRs. I did not participate in or follow the discussions due to dealing with difficult circumstances over that time period. One thing which was decided was that each factorial function should return a 0d array when given 0d input. This decision appears to have been made from first principles, and if I interpret correctly, I think @h-vetinari believes that the existing standard ufunc behavior to return a scalar when the output would otherwise be expected to be a 0d array is incorrect.

Recently, gh-18754 came to my attention, a PR for updating `factorial` to return a 0d array when given 0d input, bringing it line with the behavior of `factorial2` and `factorialk`.  This is when I first learned of the inconsistency.

I don't think it's good to have inconsistent behavior here so would like to settle down what the correct behavior should be. 

## TLDR

The standard behavior for ufuncs in `numpy` and `scipy.special` is to return a scalar when the output would otherwise be expected to be a 0d array. As part of the factorial cleanup in gh-15600, it was decided that the factorial functions would return 0d arrays when given 0d input. We need to resolve this inconsistency and settle down what behavior should be preferred.

## ?

I'd like to hear from developers who know why the current standard of converting 0d arrays to scalar upon return from ufuncs was chosen. My thought is that like Chesterton's fence, it must exist for a reason, and we need to understand what that is before removing it.

**Comment by steppi** [comment]
Some more information.

This [stackoverflow answer](https://stackoverflow.com/a/794812) goes into why NumPy scalars and 0d arrays are treated as the same thing despite technically being different types.

If I understand correctly, I think this is saying that having a 0d array is a matter of implementation convenience/necessity, and treating 0d arrays as the same thing as scalars ensures that NumPy plays nicely with pure Python. I want to try to understand more of the specifics.

**Comment by steppi** [comment]
I have found some past NumPy issues which discuss the inconsistency between 0d array and scalar.

https://github.com/numpy/numpy/issues/4563 has some discussion which began in 2014 and fizzled out without resolution in 2017.

https://github.com/numpy/numpy/issues/13105 has more discussion with a suggested path forward. There's even an associated PR, https://github.com/numpy/numpy/pull/14489 which was still active as of this past April. This allows one to signal not to convert to scalar by passing a sentinel through the `out` keyword argument.

It looks like it is generally agreed that the behavior to return convert 0d arrays to scalar before return isn't great, at least I haven't seen anyone defending it. There's still activity surrounding this on the NumPy side, and it seems feasible that in the future, 0d array types will be preserved, perhaps in a version 2.0. Could it cause trouble/confusion if the factorial functions differ from the current expected behavior to convert? 

I think my preference would be to just follow NumPy as closely as possible, and migrate towards a new behavior in synchrony with them. We've already made `0d array -> 0d array` the default behavior for the factorial functions though, and it would also be somewhat annoying to change this to `0d array -> scalar`. I would be really interested in hearing people's thoughts on this.

**Comment by h-vetinari** [comment]
Thanks for opening this RFC!

> if I interpret correctly, I think @h-vetinari believes that the existing standard ufunc behavior to return a scalar when the output would otherwise be expected to be a 0d array is incorrect.

To be clear: #15600 did not intend to break any long-standing conventions, just to make the various factorial functions consistent between each other. Aligning with `_ufuncs._factorial` was not really on my radar; partly because I was focussed on `scipy.special.factorial{,2,k}`, but also because it's a private function. Mostly, I find that ufuncs are less obviously a candidate for consistency with directly exported functions - they are designed to run in more constrained settings (e.g. only on arrays) and have some different requirements as a consequence of that.

That said, I do believe maintaining "array-ness" is desirable, and that it hasn't been done (from what I can tell) mostly due to longwinded historical reasons. Treating 0d arrays as scalars introduces a "discontinuity" in the array-ness of the output across the different dimensions, which makes programming generically for arrays really hard (which is especially important for libraries, e.g. if you need to ensure that the results still have a usable array API after whatever operation you perform).

So I wouldn't go as far as saying that the ufunc behaviour is incorrect, but certainly not ideal IMO.

> numpy/numpy#13105 has more discussion with a suggested path forward. There's even an associated PR, numpy/numpy#14489 which was still active as of this past April.

Although I haven't followed this closely, this goes in the direction of my gut feeling that the behaviour is a mix of convenience and historical cruft, and that numpy is also interested in cleaning this up, though that takes a lot of effort.

> I think my preference would be to just follow NumPy as closely as possible, and migrate towards a new behavior in synchrony with them if needed.

I'm curious to hear opinions here, but in my impression, a slow march towards cleaning up this inaccurate behaviour is not something bad (as long as we don't open up too many internal inconsistencies).

> We've already made `0d array -> 0d array` the default behavior for the factorial functions though, and it would also be somewhat annoying to change this to `0d array -> scalar`.

In this particular case, I think the situation pre-#15600 was strictly worse than what we have now, and people seemed to manage nevertheless. I bring this as supporting evidence that we could very likely "live" with the inconsistency to ufuncs, especially in light of a potential path towards cleaning this up in the future (and the inherently stricter requirements for ufuncs mentioned above).

I would prefer not to rebreak this behaviour, as it will almost inevitably hit someone in real life. That said, I'm not categorically against changing this back, and I do see the value of consistency.

**Comment by steppi** [comment]
> In this particular case, I think the situation pre-#15600 was strictly worse than what we have now, and people seemed to manage nevertheless.

Completely agreed that things have vastly improved following gh-15600. 


> I would prefer not to rebreak this behaviour, as it will almost inevitably hit someone in real life. That said, I'm not categorically against changing this back, and I do see the value of consistency.

For everyone else here, a practical consideration which hasn’t been explicitly mentioned but which explains why `0d array -> scalar` is likely to hit someone for the factorial functions when this behavior isn’t much of a concern for most ufuncs, is that the returned value can be an array of `dtype` `object` which contains Python `int`s. NumPy scalars have array methods and attributes and shouldn’t break generic array programming, but if a 0d array will be converted to a single Python `int`, you can not trust that array operations can be used on the output of a factorial function. `0d array -> scalar` is the current behavior for ufuncs which work with `object` arrays, and the problems in this case are mentioned in one of the NumPy discussions I linked above.

**Comment by rgommers** [comment]
> It looks like it is generally agreed that the behavior to convert 0d arrays to scalar before return isn't great, at least I haven't seen anyone defending it. 

Yes, that sounds right. It's not incorrect, just not great. If we could rewind 20 years of history, array scalars would not exist at all.

> There's still activity surrounding this on the NumPy side, and it seems feasible that in the future, 0d array types will be preserved, perhaps in a version 2.0. Could it cause trouble/confusion if the factorial functions differ from the current expected behavior to convert?

We discussed this for 2.0, but it's not going to happen there - primarily because it'd be quite a large amount of work, and there's no one to take that on. Also given other changes and reviewer bandwidth, it's not feasible to work on that. It may happen in a future release, or never. 



> For everyone else here, a practical consideration which hasn’t been explicitly mentioned but which explains why `0d array -> scalar` is likely to hit someone for the factorial functions when this behavior isn’t much of a concern for most ufuncs, is that the returned value can be an array of `dtype` `object` which contains Python `int`s.

Object arrays are a pain, and we should try not to return them if we can avoid it. That said, a mix of arrays and scalars depending on input values is not really better.

My input on the decision here:
- for `factorial` specifically I don't mind either way.
- In general, 0-D arrays are better than scalars, so for other modules where we don't have a lot of ufuncs that already return scalars, using 0-D arrays consistently is better.
  - This will also play better with static typing and with the efforts around support for other array libraries, which all do not have scalars.

**Comment by steppi** [comment]
> Object arrays are a pain, and we should try not to return them if we can avoid it. That said, a mix of arrays and scalars depending on input values is not really better.

I think this is one of the few situations where `object` arrays are the best option. Factorials grow so quickly that they'll rapidly overflow machine integers, `factorial(21) = 51090942171709440000`, doesn't fit in a 64 bit signed `int`. For exact calculations with the `exact=True` option, arbitrary-precision `int`s are  needed to make a useful factorial function. 

Since calculating `factorial(n)` exactly involves calculating `factorial(k)` for all natural numbers `k < n`, it's useful to be able to take an array of numbers and avoid redundant computation when computing factorials for all of them. Since we want to do the calculations with arbitrary-precision `int`s, we're stuck working with and returning an `object` array of `int`s. @rgommers, let me know if you think there's a better option. Perhaps in the future NumPy could add support for arrays of arbitrary-precision `int`s? 

> * for `factorial` specifically I don't mind either way.
> 
> * In general, 0-D arrays are better than scalars, so for other modules where we don't have a lot of ufuncs that already return scalars, using 0-D arrays consistently is better.
>   
>   * This will also play better with static typing and with the efforts around support for other array libraries, which all do not have scalars.

OK, I think 0-D arrays for factorial functions and other combinatorial functions which offer an `exact=True` option sounds reasonable then, but we should document this clearly. I think `stirling2` from gh-18103 should also be updated to use 0-D arrays then, since it works similarly to  the factorial functions.

**Comment by adeak** [comment]
> I think this is one of the few situations where `object` arrays are the best option. Factorials grow so quickly that they'll rapidly overflow machine integers, `factorial(21) = 51090942171709440000`, doesn't fit in a 64 bit signed `int`. For exact calculations with the `exact=True` option, arbitrary-precision `int`s are needed to make a useful factorial function.

As an end user, that's why I'd actually never expect such factorial functionality from an array-based library :) do you know if the benefit of (I assume) using C code instead of naive python loops with ints is significant considering the size of ints appearing for non-trivial arguments? I mean large ints are slow to work with, so perhaps the object arrayness doesn't add much in terms of computational time. There's still the obvious benefit from the array API though.

**Comment by h-vetinari** [comment]
I'm encountering this again with continued work on the `factorial*` functions, and having to work around this in various ways to ensure 0dim array inputs are not "degraded" to scalars when passing through (e.g.) the gamma function. Would be nice to fix this more comprehensively.

**Comment by steppi** [comment]
> I'm encountering this again with continued work on the `factorial*` functions, and having to work around this in various ways to ensure 0dim array inputs are not "degraded" to scalars when passing through (e.g.) the gamma function. Would be nice to fix this more comprehensively.

I agree. It's going to be difficult though. This behavior of having 0dim arrays converted to scalars is baked into the NumPy C API. 

rgommers had said:

> We discussed this for 2.0, but it's not going to happen there - primarily because it'd be quite a large amount of work, and there's no one to take that on. Also given other changes and reviewer bandwidth, it's not feasible to work on that. It may happen in a future release, or never.

@h-vetinari, I think you'd to champion this to make it happen, and it will be a long road.
